### DISCUSSION TITLE
Middleware return 500 on Production

### DISCUSSION THREAD

**Discussion by yongsk0066** [vercel/next.js#33636] [Help]
### Run `next info` (available from version 12.0.8 and up)

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64
    Binaries:
      Node: 14.18.1
      npm: 6.14.15
      Yarn: 1.22.10
      pnpm: N/A
    Relevant packages:
      next: 12.0.8
      react: 17.0.2
      react-dom: 17.0.2

### What version of Next.js are you using?

12.0.8

### What version of Node.js are you using?

14.18.1

### What browser are you using?

chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next start on aws EC2 instance by container

### Describe the Bug

pages/_middleware.ts
```Typescript
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

import { getAuthAPI } from 'envs';
import { joinDictionary } from 'utils/util';

const UNAUTHORIZED = 'require_authorize';
const AUTHORIZED = 'ok';

export async function middleware(request: NextRequest) {
  const res = await fetch(`${getAuthAPI()}/checkAuth`, {
    method: 'GET',
    headers: {
      cookie: joinDictionary(request.cookies),
    },
  }).then((res) => res.json());
  const redirectUrl = request.nextUrl.searchParams.get('redirect_url');

  if (res.code === UNAUTHORIZED && request.nextUrl.pathname !== '/login') {
    return NextResponse.redirect(`/login?redirect_url=${request.url}`);
  }
  if (res.code === AUTHORIZED && request.nextUrl.pathname === '/login') {
    return NextResponse.redirect(redirectUrl ? redirectUrl : '/');
  }
  return NextResponse.next();
}
```

I wrote middleware on top of the pages dir to check user is authorized on every pages.
It works really well on local but if i deploy on EC2 or fargate, then it returns 500 error.

I tried 'curl "http://localhost/' to check health likes ec2 targetGroup healthchecking on both local and inside of EC2 instance.
local returns 200 but on EC2 returns 500. 

### Expected Behavior

return StatusCode 200

### To Reproduce

excute curl "http://localhost/" to check health

**Comment by aar2dee2** [comment]
Hi,

I'm facing the same error when using middleware to set up subdomains in my vercel app. Did you find a fix for this?

```ts
import { NextResponse } from "next/server";

import type { NextRequest } from "next/server";
//code for this page was taken from here https://github.com/vercel/platforms/blob/main/pages/_middleware.ts
// and this post from vercel was referenced: https://vercel.com/guides/nextjs-multi-tenant-application
export default function middleware(req: NextRequest) {
  console.log("url is", req.nextUrl)
  console.log("headers are", req.headers)
  // Clone the request url
  const url = req.nextUrl.clone();

  // Get pathname of request (e.g. /blog-slug)
  const { pathname } = req.nextUrl;

  // Get hostname of request (e.g. demo.vercel.pub)
  const hostname = req.headers.get("host");
  console.log("extracted hostname is", hostname)
  if (!hostname)
    return new Response(null, {
      status: 400,
      statusText: "No hostname found in request headers",
    });

  const currentHost =
    process.env.NODE_ENV === "production" //&& process.env.VERCEL === "1"
      ? // You have to replace ".vercel.pub" with your own domain if you deploy this example under your domain.
        // You can use wildcard subdomains on .vercel.app links that are associated with your Vercel team slug
        // in this case, our team slug is "platformize", thus *.platformize.vercel.app works
        hostname
          .replace(`.telborg.com`, "")
          .replace(`.telborg.vercel.app`, "")
      : hostname.replace(`.localhost:3000`, "");
  console.log("node env is", process.env.NODE_ENV)
  console.log("vercel num is", process.env.VERCEL)
  console.log("current host is", currentHost)
  if (pathname.startsWith(`/_sites`))
    return new Response(null, {
      status: 404,
    });

  if (!pathname.includes(".") && !pathname.startsWith("/api")) {
    if (currentHost == "app") {
      if (
        pathname === "/login" &&
        (req.cookies["next-auth.session-token"] ||
          req.cookies["__Secure-next-auth.session-token"])
      ) {
        url.pathname = "/";
        return NextResponse.redirect(url);
      }

      url.pathname = `/app${pathname}`;
      return NextResponse.rewrite(url);
    }

    if (
      hostname === "localhost:3000" ||
      hostname === "telborg.com"
    ) {
      url.pathname = `/home${pathname}`;
      return NextResponse.rewrite(url);
    }

    url.pathname = `/_sites/${currentHost}${pathname}`;
    return NextResponse.rewrite(url);
  }
}
```

**Reply by icyJoseph** [reply]
Do you have more info/logs or repo to look at?

**Reply by aar2dee2** [reply]
@icyJoseph 
Yup, 

From the Functions log on Vercel:
<img width="1143" alt="Screenshot 2022-05-02 at 2 14 09 PM" src="https://user-images.githubusercontent.com/85004512/166208177-2b31136d-a2dc-4326-bd7d-dbd6ed2520b5.png">

Deployment Status:
<img width="1090" alt="Screenshot 2022-05-02 at 2 15 02 PM" src="https://user-images.githubusercontent.com/85004512/166208399-8f4895f6-76bd-40f0-9900-a007fa421bcf.png">

Checkly:
<img width="1347" alt="Screenshot 2022-05-02 at 2 15 28 PM" src="https://user-images.githubusercontent.com/85004512/166208433-95d20678-fe7f-4891-85d0-0ab42f4aa8e2.png">

**Reply by aar2dee2** [reply]
I've set up a subdomain `app.telborg.com`. In development, the routing works correctly - `app.localhost:3000` resolves to `/pages/app` directory.

**Reply by icyJoseph** [reply]
Interesting that error is a React invariant breakage. Are you sure the right data is resolving when trying to render the page?

https://reactjs.org/docs/error-decoder.html/?invariant=31&args=%5B%5D%3Dobject%20with%20keys%20%7Bchildren%7D

**Reply by aar2dee2** [reply]
Thanks for pointing that out!
Could there be an issue with the `_app.tsx`? I'm using two different layouts to serve the main domain and the `app.telborg.com` subdomain.

This is my `_app.tsx`
```ts
//import from packages
import '../styles/globals.css'
import { config } from '@fortawesome/fontawesome-svg-core'
import '@fortawesome/fontawesome-svg-core/styles.css'
import type { AppProps } from 'next/app'
//import from context and utils
import { UserProvider } from '../context/user'
//import from components
import AppLayout from '../components/app/appLayout'
import HomeLayout from '../components/home/homeLayout'
import { NextPageContext } from 'next'

interface CustomAppProps extends Omit<AppProps, "Component"> {
  Component: AppProps["Component"] & { layout: "string" };
}

const layouts = {
  AppL: AppLayout,
  HomeL: HomeLayout, 
}

config.autoAddCss = false;

function MyApp({ Component, pageProps }: CustomAppProps) {
  const selected = Component.layout;
  const Layout = layouts[selected as keyof typeof layouts] || ((children) => <>{children}</>);
  return (
  <UserProvider>
  <Layout >
    <linearGradient id="icon-gradient" x2="0" y2="1">
        <stop offset="0%" stopColor="#C3F73A" />
        <stop offset="50%" stopColor="#B5B68D" />
        <stop offset="100%" stopColor="#A360FB" />
    </linearGradient>
    <Component {...pageProps} />
  </Layout>
  </UserProvider>)
}

export default MyApp
```

and the `homeLayout.tsx`
```ts
//import from packages
import Head from "next/head"
import React from "react"
//import from context and utils
//import from components
import HomeNavbar from "./homeNavbar"
//import assets

interface AppProps {

}
export default function HomeLayout({ children }: React.PropsWithChildren<AppProps>) {
    return (
        <>
            <Head>
                <title>Telborg</title>
                <link rel="icon" href="/favicon.ico" ></link>
                <meta charSet="utf-8" />
                <meta name="viewport" content="initial-scale=1.0, width=device-width" />
            </Head>
            <header>
                <HomeNavbar />
            </header>
            <main className="pt-48">{children}</main>
            {/* <Footer /> */}
        </>
    )
}
```

I pass the `.layout` prop as a string to each page separately. For example, the home page is as below.
```ts
//import from packages
//import from context and utils
//import from components
import Hero from "../../components/Hero"

export default function Home() {
    return (
        <>
            <Hero />
        </>
    )
}

Home.layout = "HomeL";
```

**Reply by icyJoseph** [reply]
I think you have a bit of weird per page layout system, compared with the suggested in the docs anyway, but there's a typo here:

```js
 const Layout = layouts[selected as keyof typeof layouts] || ((children) => <>{children}</>);
```

The way you have it, it should be:

```js
 const Layout = layouts[selected as keyof typeof layouts] || (({ children }) => <>{children}</>);
```

Notice that the [docs](https://nextjs.org/docs/basic-features/layouts#per-page-layouts) do:

```js
// Use the layout defined at the page level, if available
  const getLayout = Component.getLayout ?? ((page) => page) // no fragment
```

**Reply by aar2dee2** [reply]
Thanks again. I corrected the typo. (I think that fixed the React error 🙏🏼).
I think with the layouts - I'm importing the layout component in only `_app`, rather than every page, and passing a `layout` prop to select the correct one.

So, after adding `_error` and `400` and `500` routes, I'm no longer getting a `500` server error, but a `page not found: 404`.

checkly: 
<img width="1339" alt="Screenshot 2022-05-02 at 5 09 26 PM" src="https://user-images.githubusercontent.com/85004512/166227902-8d31cba1-fc6e-4562-8fa1-494adb3aa681.png">
<img width="1349" alt="Screenshot 2022-05-02 at 5 09 42 PM" src="https://user-images.githubusercontent.com/85004512/166227932-52cb8dd1-852c-4e13-8d62-c7defcae1585.png">

No errors in the Logflare logs for the latest deployment.

**Reply by icyJoseph** [reply]
What's up with the `user details are null` - I wonder if a header you are expecting is not coming through to the server. Do the middleware logs show what you expect?

**Reply by aar2dee2** [reply]
The `user details are null` was just a console.log I'd done to check the layouts were rendering correctly.

I tried removing the `_middleware.ts` file. Without that, the deployment succeeds and all paths render correctly. For instance, those in app  render under `/app/<route>`.  For writing the code for the `_middleware` file, I referred [this blog from Vercel](https://vercel.com/guides/nextjs-multi-tenant-application#4.-configure-rewrites-for-multi-tenancy) and this [corresponding code file](https://github.com/vercel/platforms/blob/5d2479f525b3f2611af7215e7b3465aa80f6d2a5/pages/_middleware.ts).

Here's my `_middleware.ts`:
```ts
import { NextResponse } from "next/server";

import type { NextRequest } from "next/server";
//code for this page was taken from here https://github.com/vercel/platforms/blob/main/pages/_middleware.ts
// and this post from vercel was referenced: https://vercel.com/guides/nextjs-multi-tenant-application
export default function middleware(req: NextRequest) {
  // Clone the request url
  const url = req.nextUrl.clone();

  // Get pathname of request (e.g. /blog-slug)
  const { pathname } = req.nextUrl;

  // Get hostname of request (e.g. demo.vercel.pub)
  const hostname = req.headers.get("host");

  if (!hostname)
    return new Response(null, {
      status: 400,
      statusText: "No hostname found in request headers",
    });

  const currentHost =
    process.env.NODE_ENV === "production" && process.env.VERCEL === "1"
      ? // You have to replace ".vercel.pub" with your own domain if you deploy this example under your domain.
        // You can use wildcard subdomains on .vercel.app links that are associated with your Vercel team slug
        // in this case, our team slug is "platformize", thus *.platformize.vercel.app works
        hostname
          .replace(`www.telborg.com`, "")
          .replace(`.telborg.com`, "")
          .replace(`.telborg.vercel.app`, "")
      : hostname.replace(`.localhost:3000`, "");

  if (pathname.startsWith(`/_sites`))
    return new Response(null, {
      status: 404,
    });

  if (!pathname.includes(".") && !pathname.startsWith("/api")) {
    if (currentHost == "app") {
      if (
        pathname === "/login" &&
        (req.cookies["next-auth.session-token"] ||
          req.cookies["__Secure-next-auth.session-token"])
      ) {
        url.pathname = "/app";
        return NextResponse.redirect(url);
      }

      url.pathname = `/app${pathname}`;
      return NextResponse.rewrite(url);
    }

    if (
      hostname === "localhost:3000" ||
      hostname === "www.telborg.com"
    ) {
      url.pathname = `/home${pathname}`;
      return NextResponse.rewrite(url);
    }

    url.pathname = `/_sites/${currentHost}${pathname}`;
    return NextResponse.rewrite(url);
  }
}
```

**Reply by aar2dee2** [reply]
this is from logflare:
<img width="730" alt="Screenshot 2022-05-02 at 8 56 37 PM" src="https://user-images.githubusercontent.com/85004512/166261170-3c7e07fe-5c7a-427e-be2e-c4ae1b3943bc.png">

from checkly:
<img width="1226" alt="Screenshot 2022-05-02 at 8 58 06 PM" src="https://user-images.githubusercontent.com/85004512/166261380-bfc00939-c3de-46c8-b919-85f726cdd680.png">

**Reply by icyJoseph** [reply]
In the logflare log, host is `telborg-foo-bar.vercel.app`,  and that's not going to make `currentHost` equal to "app". 

That's looking like a bug to me, could you double check it?

Also, if possible, add a status text to this block:

```js
if (pathname.startsWith(`/_sites`))
    return new Response(null, {
        status: 404,
        statusText: '_sites not found via direct access',
    });

```

**Reply by aar2dee2** [reply]
the host `telborg-foo-bar.vercel.app` is the url assigned to that deployment by vercel. It should resolve to `""` - the main domain, not the `app` subdomain. I reworked the `currentHost` code as follows:
```ts
const currentHost = (() => {
    return (
      (process.env.NODE_ENV === "production" && process.env.VERCEL === "1") ? 
      hostname
      .replace(`www.telborg.com`, "")
      .replace(`.telborg.com`, "")
      .replace(`.telborg.vercel.app`, "")
      : (hostname.includes(".vercel.app")) ? hostname.replace(`.${process.env.VERCEL_URL}`, "")
      : hostname.replace(`.localhost:3000`, ""));
  })();
```

I added the `statusText` as suggested. Here's the latest 404 from Logflare:
<img width="795" alt="Screenshot 2022-05-03 at 12 26 17 AM" src="https://user-images.githubusercontent.com/85004512/166307972-c854389d-a55a-42ff-a4de-77c621538ae2.png">

**Reply by icyJoseph** [reply]
This is very weird. The message in the logflare logs are the output of running `next build`. It's saying that a 404 static page has been generated. Nothing more.

Between the file you used as reference and your code I only see one major difference:

```js
 if (currentHost == "app") {
      if (
        pathname === "/login" &&
        (req.cookies["next-auth.session-token"] ||
          req.cookies["__Secure-next-auth.session-token"])
      ) {
        url.pathname = "/"; // HERE
        return NextResponse.redirect(url);
      }

      url.pathname = `/app${pathname}`;
      return NextResponse.rewrite(url);
    }
```

I remember implementing this https://github.com/vercel/examples/blob/main/edge-functions/hostname-rewrites/pages/_middleware.ts example a couple of months ago and it worked. 

I'll give it a go tomorrow.

**Reply by aar2dee2** [reply]
That must be the static page for `404.tsx`. I had to include a `_error.tsx` and custom pages for `500.tsx` and `404.tsx`. This is the code for the `404.tsx`
```ts
export default function Custom404() {
    return <h1>404 - Page Not Found</h1>
  }

Custom404.layout = "HomeL"
```

These are the logs from the last deployment (which is throwing the 404 error on Checkly, but no errors on Logflare).
<img width="1030" alt="Screenshot 2022-05-03 at 12 09 14 PM" src="https://user-images.githubusercontent.com/85004512/166413442-425d7f5f-b765-4175-adb7-1cc3265e6e42.png">
<img width="795" alt="Screenshot 2022-05-03 at 12 26 17 AM" src="https://user-images.githubusercontent.com/85004512/166413500-983566d2-3d08-4134-8bfb-2a7e234e2063.png">

I removed the `_error`, `404`, `500` and the `_document` pages.  This is the log from Checkly now. There is no error with respect to `HEAD` that was appearing earlier. 
<img width="1297" alt="Screenshot 2022-05-03 at 12 51 05 PM" src="https://user-images.githubusercontent.com/85004512/166417028-d36a5439-5c39-48d2-ba31-5fb12f238b7c.png">

**Reply by icyJoseph** [reply]
Mmm it must be that something is hiding in plain sight, or the architecture is just not working as we think it should. Do you have logs from the _middleware runtime? You had placed some in your original submission.

**Reply by aar2dee2** [reply]
The middleware function is not throwing up any errors now, so no logs from that. Those logs were from when I was getting the `500` error.

**Reply by aar2dee2** [reply]
Could it be related to accessing the correct host during the deployment? I'm using the variable `process.env.VERCEL_URL` from [here](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) to access the host for the deployment in `currentHost` in `_middleware`.

```ts
const currentHost = (() => {
    return (
      (process.env.NODE_ENV === "production" && process.env.VERCEL === "1") ? 
      hostname
      .replace(`www.telborg.com`, "")
      .replace(`.telborg.com`, "")
      .replace(`.telborg.vercel.app`, "")
      .replace(`.${process.env.VERCEL_URL}`, "")
      : (hostname.includes(".vercel.app")) ? hostname.replace(`.${process.env.VERCEL_URL}`, "")
      : hostname.replace(`.localhost:3000`, ""));
```

**Reply by aar2dee2** [reply]
Also tried with `process.env.NEXT_PUBLIC_VERCEL_URL`.

**Reply by aar2dee2** [reply]
So I tried skipping the checks and promoting the failed deployment to production. 
All the paths on the `app.telborg.com` subdomain resolve correctly. You can see here [login](https://app.telborg.com/login). 
Also, all routes on `wwww.telborg.com` load as well. See [www.telborg.com](https://www.telborg.com/).

However, routes on `telborg.com` did not load. 

I added `hostname === telborg.com` to the last if clause in `_middleware` but that does not resolve it.
```ts
if (
      hostname === "localhost:3000" ||
      hostname === "www.telborg.com" ||
      hostname === "telborg.com"
    ) {
      url.pathname = `/home${pathname}`;
      return NextResponse.rewrite(url);
    }
```

This fixed the loading for routes on `telborg.com`. But the checks still throw an error.

**Reply by aar2dee2** [reply]
Update: Adding the deployment url to that if clause fixed the checks too.

```ts
if (
      hostname === "localhost:3000" ||
      hostname === "www.telborg.com" ||
      hostname === "telborg.com" ||
      hostname === process.env.NEXT_PUBLIC_VERCEL_URL
    ) {
      url.pathname = `/home${pathname}`;
      return NextResponse.rewrite(url);
    }
```

Thanks a ton for helping me out. You're awesome!! 🙏🏼

**Reply by icyJoseph** [reply]
Wait, so it works alright now? Great!

**Reply by icyJoseph** [reply]
Nice, I'd recommend writing tests for this middleware, perhaps the error would've popped earlier.

You can use `ts-mockito` to mock the NextRequest input to the middleware and check that the middleware returns rewrites(x-middleware-rewrite header presence), redirects (status code 307), that it just does a next call (response 200), that it errors with 404, or 500, or that it does nothing, by returning undefined.

```ts
import {NextURL} from 'next/dist/server/web/next-url';
// eslint-disable-next-line @next/next/no-server-import-in-page
import {NextRequest} from 'next/server'; // it's ok to import next/server here
import {instance, mock, reset, when} from 'ts-mockito';

const mockedRequest: NextRequest = mock(NextRequest);

afterEach(() => {
    reset(mockedRequest);
});

// and then test
test('Test Case', async () => {
    const url = new URL('/', 'https://teleborg.com');

    const nextUrl = new NextURL(url);

    when(mockedRequest.nextUrl).thenReturn(nextUrl);

    const result = await middleware(instance(mockedRequest));

    expect(result).toBeDefined();

    expect(result.status).toBe(200);

    expect(result.headers.get('x-middleware-rewrite')).toBe('some-expect-string');
});
```

**Reply by aar2dee2** [reply]
I'm going to do this! Thanks for sharing how to go about it.

**Reply by icyJoseph** [reply]
Sorry for reviving and old issue, but could you mark this as answered, I couldn't find this answer, and I needed it today 😅  (was searching with answered flag set to true)

**Reply by aar2dee2** [reply]
Does the person who started the discussion need to do that?

**Reply by icyJoseph** [reply]
Ah yeah, nvm, I though it was your discussion, never mind then! Have a good day, hope that middleware is behaving!

**Reply by aar2dee2** [reply]
yea, all good with the middleware :D
